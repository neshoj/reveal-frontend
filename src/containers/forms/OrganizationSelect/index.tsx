import reducerRegistry from '@onaio/redux-reducer-registry';
import { keyBy, values } from 'lodash';
import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import Select from 'react-select';
import { Store } from 'redux';
import { OPENSRP_ORGANIZATION_ENDPOINT } from '../../../constants';
import { OpenSRPService } from '../../../services/opensrp';
import store from '../../../store';
import assignmentReducer, {
  Assignment,
  fetchAssignments,
  getAssignmentsArrayByPlanId,
  reducerName as assignmentReducerName,
  resetPlanAssignments,
} from '../../../store/ducks/opensrp/assignments';
import organizationsReducer, {
  fetchOrganizations,
  getOrganizationsArray,
  Organization,
  reducerName as organizationsReducerName,
} from '../../../store/ducks/opensrp/organizations';

reducerRegistry.register(assignmentReducerName, assignmentReducer);
reducerRegistry.register(organizationsReducerName, organizationsReducer);

/** react-select Option */
interface SelectOption {
  label: string;
  value: string;
}

/** OrganizationSelect props */
export interface OrganizationSelectProps {
  assignments: Assignment[];
  fetchAssignmentsAction: typeof fetchAssignments;
  fetchOrganizationsAction: typeof fetchOrganizations;
  jurisdictionId: string;
  name: string; // name of the input
  organizations: Organization[];
  planId: string;
  resetPlanAssignmentsAction: typeof resetPlanAssignments;
  serviceClass: typeof OpenSRPService /** the OpenSRP service */;
  value: SelectOption[];
}

/** default props for OrganizationSelect */
const defaultProps: OrganizationSelectProps = {
  assignments: [],
  fetchAssignmentsAction: fetchAssignments,
  fetchOrganizationsAction: fetchOrganizations,
  jurisdictionId: '',
  name: '',
  organizations: [],
  planId: '',
  resetPlanAssignmentsAction: resetPlanAssignments,
  serviceClass: OpenSRPService,
  value: [],
};

/**
 * OrganizationSelect - a flat select for Organizations
 * Allows you to Select Organizations to be assigned to the plan-jurisdiction
 * On selection update the `handleChange` method updates the Assignments store
 */
const OrganizationSelect = (props: OrganizationSelectProps) => {
  const {
    assignments,
    fetchAssignmentsAction,
    fetchOrganizationsAction,
    jurisdictionId,
    name,
    organizations,
    planId,
    resetPlanAssignmentsAction,
    serviceClass,
    value: selectOptions,
  } = props;

  const loadOrganizations = async (service: typeof serviceClass) => {
    const serve = new service(OPENSRP_ORGANIZATION_ENDPOINT);
    serve
      .list()
      .then((response: Organization[]) => store.dispatch(fetchOrganizationsAction(response)))
      .catch((err: Error) => {
        /** TODO - find something to do with error */
      });
  };

  useEffect(() => {
    loadOrganizations(serviceClass);
  }, []);

  /** Get select options from OpenSRP as a promise */

  /**
   * onChange callback
   */
  const handleChange = (nextValues: any) => {
    const filteredAssignments: Assignment[] = assignments.filter(
      (a: Assignment) => a.jurisdiction !== jurisdictionId
    );
    if (!nextValues) {
      resetPlanAssignmentsAction({ [planId]: filteredAssignments });
    } else {
      const newAssignments: Assignment[] = (nextValues || []).map(
        (v: SelectOption) =>
          ({
            jurisdiction: jurisdictionId,
            organization: v.value,
            plan: planId,
          } as Assignment)
      );
      const nextAssignments: Assignment[] = [...filteredAssignments, ...newAssignments];
      fetchAssignmentsAction(nextAssignments);
    }
  };
  const options = organizations.map(
    o =>
      ({
        label: o.name,
        value: o.identifier,
      } as SelectOption)
  );

  return (
    <Select
      name={name}
      bsSize="lg"
      placeholder={'Select'}
      className="organization-select"
      classNamePrefix="reveal"
      aria-label={'Select'}
      onChange={handleChange}
      defaultOptions={true}
      options={options}
      isClearable={true}
      isMulti={true}
      value={selectOptions}
    />
  );
};

OrganizationSelect.defaultProps = defaultProps;

export { OrganizationSelect };

// connect to store
const mapStateToProps = (state: Partial<Store>, ownProps: OrganizationSelectProps) => {
  const organizations = keyBy(getOrganizationsArray(state), (o: Organization) => o.identifier);
  const assignments = getAssignmentsArrayByPlanId(state, ownProps.planId);
  const selectOptions = assignments
    .filter((a: Assignment) => a.jurisdiction === ownProps.jurisdictionId)
    .map(
      (a: Assignment) =>
        ({
          label:
            (organizations[a.organization] && organizations[a.organization].name) || a.organization,
          value: a.organization,
        } as SelectOption)
    );
  return {
    ...ownProps,
    assignments,
    organizations: values(organizations),
    value: selectOptions,
  };
};

/** map props to actions that may be dispatched by component */
const mapDispatchToProps = {
  fetchAssignmentsAction: fetchAssignments,
  resetPlanAssignmentsAction: resetPlanAssignments,
};

const ConnectedOrganizationSelect = connect(
  mapStateToProps,
  mapDispatchToProps
)(OrganizationSelect);

export default ConnectedOrganizationSelect;
