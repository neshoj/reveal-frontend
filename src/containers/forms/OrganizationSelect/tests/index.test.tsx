import reducerRegistry from '@onaio/redux-reducer-registry';
import { mount, shallow } from 'enzyme';
import React from 'react';
import { Provider } from 'react-redux';
import OrganizationSelect from '..';
import store from '../../../../store';
import assignmentReducer, {
  reducerName as assignmentReducerName,
} from '../../../../store/ducks/opensrp/assignments';
import organizationsReducer, {
  fetchOrganizations,
  reducerName as organizationsReducerName,
} from '../../../../store/ducks/opensrp/organizations';
import { organizations } from '../../../../store/ducks/tests/fixtures';

jest.mock('../../../../configs/env');

reducerRegistry.register(assignmentReducerName, assignmentReducer);
reducerRegistry.register(organizationsReducerName, organizationsReducer);

describe('/containers/forms/OrganizationSelect', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('renders without crashing', () => {
    shallow(<OrganizationSelect />);

    mount(
      <Provider store={store}>
        <OrganizationSelect jurisdictionId="" name="" planId="" />
      </Provider>
    );
  });

  it('renders select options correctly', () => {
    store.dispatch(fetchOrganizations(organizations));

    const props = {
      jurisdictionId: 'outpost-number-one',
      name: 'plan-assignment-form-outpost-number-one',
      planId: 'alpha',
    };

    const wrapper = mount(
      <Provider store={store}>
        <OrganizationSelect {...props} />
      </Provider>
    );

    expect(wrapper.children().props()).toMatchSnapshot('Organization Select ownProps');
    expect(
      wrapper
        .children()
        .children()
        .props()
    ).toMatchSnapshot('Organization Select Props');
    expect(wrapper.find('.reveal__indicator').length).toBe(2);
  });
});
