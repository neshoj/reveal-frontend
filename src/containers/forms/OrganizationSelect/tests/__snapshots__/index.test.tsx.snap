// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/containers/forms/OrganizationSelect renders select options correctly: Organization Select Props 1`] = `
Object {
  "assignments": Array [],
  "fetchAssignmentsAction": [Function],
  "fetchOrganizationsAction": [Function],
  "jurisdictionId": "outpost-number-one",
  "name": "plan-assignment-form-outpost-number-one",
  "organizations": Array [
    Object {
      "active": true,
      "id": 1,
      "identifier": "fcc19470-d599-11e9-bb65-2a2ae2dbcce4",
      "name": "The Luang",
      "type": Object {
        "coding": Array [
          Object {
            "code": "team",
            "display": "Team",
            "system": "http://terminology.hl7.org/CodeSystem/organization-type",
          },
        ],
      },
    },
    Object {
      "active": true,
      "id": 3,
      "identifier": "4c506c98-d3a9-11e9-bb65-2a2ae2dbcce4",
      "name": "Demo Team",
    },
  ],
  "planId": "alpha",
  "resetPlanAssignmentsAction": [Function],
  "serviceClass": [Function],
  "value": Array [],
}
`;

exports[`/containers/forms/OrganizationSelect renders select options correctly: Organization Select ownProps 1`] = `
Object {
  "jurisdictionId": "outpost-number-one",
  "name": "plan-assignment-form-outpost-number-one",
  "planId": "alpha",
}
`;
